src/rust/engine/ui/src/console_ui.rs

```diff
@@ -78,18 +77,36 @@ impl ConsoleUI {
   ///
   /// Setup progress bars, and return them along with a running task that will drive them.
   ///
+  /// NB: This method must be very careful to avoid logging. Between the point where we have taken
+  /// exclusive access to the Console and when the UI has actually been initialized, attempts to
+  /// log from this method would deadlock (by causing the method to wait for _itself_ to finish).
+  ///
   fn setup_bars(
     &self,
     executor: Executor,
   ) -> Result<(Vec<ProgressBar>, MultiProgressTask), String> {
-    // Stderr is propagated across a channel to remove lock interleavings between stdio and the UI.
-    let (stderr_sender, stderr_receiver) = mpsc::sync_channel(0);
-    let (term_read, _, term_stderr_write) =
+    // We take exclusive access to stdio by registering a callback that is used to render stderr
+    // while we're holding access. See the method doc.
+    let stderr_dest_bar: Arc<Mutex<Option<WeakProgressBar>>> = Arc::new(Mutex::new(None));
+    // We acquire the lock before taking exclusive access, and don't release it until after
+    // initialization. That way, the exclusive callback can always assume that the destination
+    // is initialized (i.e., can `unwrap` it).
+    let mut stderr_dest_bar_guard = stderr_dest_bar.lock();
+    let (term_read, _, term_stderr_write) = {
+      let stderr_dest_bar = stderr_dest_bar.clone();
       stdio::get_destination().exclusive_start(Box::new(move |msg: &str| {
-        // If we fail to send, it's because the UI has shut down: we fail the callback to
-        // have the logging module directly log to stderr at that point.
-        stderr_sender.send(msg.to_owned()).map_err(|_| ())
-      }))?;
+        // Acquire a handle to the destination bar in the UI. If we fail to upgrade, it's because
+        // the UI has shut down: we fail the callback to have the logging module directly log to
+        // stderr at that point.
+        let dest_bar = {
+          let stderr_dest_bar = stderr_dest_bar.lock();
+          // We can safely unwrap here because the Mutex is held until the bar is initialized.
+          stderr_dest_bar.as_ref().unwrap().upgrade().ok_or(())?
```

Comment:
This change appears to work, but I also never see a failure to upgrade the progress bar from weak; i.e.: I neve see a log::warn / retry.

Reply:
When the `stderr` callback fails, the `stdio` module will clear exclusive access and fall back to directly accessing stdio: the empty error is the signal to fallback. So that's expected!

Reply:
My read of the stdio crate is that it should fall back to looging and that should include a warn level log lione I do not see:
https://github.com/pantsbuild/pants/blob/2b909b214da7d20b8e94d9ca63966483e04dfe25/src/rust/engine/stdio/src/lib.rs#L340-L357

Reply:
Hm. Actually, you're right. When we fail to write to the handler and [tear down `stderr`](https://github.com/pantsbuild/pants/blob/0aabd417a772ea4e39999c4415c67db40de679a4/src/rust/engine/stdio/src/lib.rs#L315-L359), we _also_ clear the `Console`, which would cause things to be logged to `.pants.d/pants.log`, rather than to stderr.

So... things should probably work the way I was claiming above (an error in the stderr callback should clear exclusive access and re-enable the `Console`), but they don't currently. You can check the log to see if you're getting any output there, but I suspect that removing an extra thread and channel here might also have made the error harder to reproduce.

Reply:
Only 1 hit and that was back on 12/22 - i.e.: not from this fix:
```
$ grep -r "Failed to write" .pants.d/
.pants.d/run-tracker/pants_run_2021_12_22_11_37_03_636_20047a7c6a314cd6876147aeee73d2b7/logs:11:47:27.31 [WARN] Failed to write stderr to Console { console: Console { stdin_handle: Some(File { fd: 163, path: "/dev/pts/5 (deleted)", read: true, write: false }), stdout_handle: Some(File { fd: 164, path: "/dev/pts/5 (deleted)", read: false, write: true }), stderr_handle: Some(File { fd: 313, path: "/dev/pts/5 (deleted)", read: false, write: true }), stderr_use_color: true } }, falling back to Logging: Err("Input/output error (os error 5)")
```